components/brave_wallet/browser/ethereum_provider_impl.cc

```diff
@@ -624,33 +711,34 @@ void EthereumProviderImpl::SignTypedMessage(
     return RejectInvalidParams(std::move(id), std::move(callback));
   }
 
-  SignMessageInternal(
-      EthAddress::FromHex(address).ToChecksumAddress(), domain_string, message,
-      std::move(*message_to_sign), base::HexEncode(domain_hash),
-      base::HexEncode(primary_hash), true, std::move(callback), std::move(id));
+  mojom::SignDataUnionPtr sign_data =
+      mojom::SignDataUnion::NewEthSignTypedData(mojom::EthSignTypedData::New(
+          message, domain_string, base::HexEncode(domain_hash),
+          base::HexEncode(primary_hash)));
+
+  SignMessageInternal(EthAddress::FromHex(address).ToChecksumAddress(),
+                      std::move(sign_data), std::move(*message_to_sign),
+                      std::move(callback), std::move(id));
 }
 
 void EthereumProviderImpl::SignMessageInternal(
     const std::string& address,
-    const std::string& domain,
-    const std::string& message,
+    mojom::SignDataUnionPtr sign_data,
     std::vector<uint8_t>&& message_to_sign,
-    const absl::optional<std::string>& domain_hash,
-    const absl::optional<std::string>& primary_hash,
-    bool is_eip712,
     RequestCallback callback,
     base::Value id) {
+  CHECK(sign_data);
```

Comment:
Can we double check the memory usage of `sign_data` here? Seems like that move on line 740 below could lead to memory corruption issues if this check fails. 

At least that's what it looks like if I'm understanding this correctly.

Reply:
?! If it is nullptr, check will fail and it will intentionally crash here without moving forward.

Reply:
What's the benefit to intentional crashing rather than error handling here?

Reply:
There is no error to be handled that internal function should not be called with nullptr as sign_data
https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/checks.md#check_dcheck_notreached_noreturn_and-notreached

Reply:
Ahh ok, thanks for the link. I was wondering about this after asking that questions so super useful to understand the purpose of these better.